esp/services/ws_machine/ws_machineService.cpp

```diff
@@ -180,6 +177,11 @@ void Cws_machineEx::init(IPropertyTree *cfg, const char *process, const char *se
 
     Owned<IComponentStatusFactory> factory = getComponentStatusFactory();
     factory->init(pServiceNode);
+
+    unsigned machineUsageCacheForceRebuildMinutes = pServiceNode->getPropInt("MachineUsageCacheMinutes", MACHINE_USAGE_CACHE_MINUTES);
+    unsigned machineUsageCacheAutoRebuildMinutes = pServiceNode->getPropInt("MachineUsageCacheAutoRebuildMinutes", DEFAULT_MACHINE_USAGE_CACHE_AUTO_BUILD_MINUTES);
+    usageCacheReader.setown(new CUsageCacheReader(this));
+    usageCacheReaderThread.setown(new CInfoCacheReaderThread(usageCacheReader, "Usage Reader", machineUsageCacheAutoRebuildMinutes*60, machineUsageCacheForceRebuildMinutes*60));
```

Comment:
I think ok as it is, but logically the callers of this should be dealing with a reader interface (not a "thread"), and the builder thread (CInfoCacheReaderThread) should be contained with in it (e.g. contained within CUsageCacheReader), with the reader thread interface looking like:
```
interface IInfoCacheReader
{
    virtual CInfoCache *buildCachedInfo() = 0;
    virtual CInfoCache *getCachedInfo() = 0;
    virtual void setActive(bool b) = 0; // avoid use of "detached" which doesn't mean very much to this impl itself.
    virtual bool isActive() = 0;
};
```

then the thread contained thread CInfoCacheReaderThread can call back to buildCacheInfo(), and others can to if immediate up-to-date result is required.
So construction code would look something like this:
```
    usageCacheReader.setown(new CUsageCacheReader(this, machineUsageCacheAutoRebuildMinutes, machineUsageCacheForceRebuildMinutes));
```
CUsageCacheReader would implement the IInfoCacheReader interface and call through to it's contained thread if necessary.

Not sure it's worth refactoring, but something to consider.

Reply:
Both CUsageCacheReader and its contained thread need to access those variables (first, firstBlacked, active, sem, crit, infoCache, and waiting). Not sure it's worth refactoring like that.